libcaf_net/caf/net/web_socket/framing.cpp

```diff
@@ -63,119 +78,167 @@ void framing::abort(const error& reason) {
+  if (hdr_.fin) {
+    if (opcode_ == detail::rfc6455::opcode_type::nil_code) {
+      if (hdr_.opcode == detail::rfc6455::opcode_type::text_frame
+          && validation_offset_ != payload.size()) {
         abort_and_shutdown(sec::malformed_message, "invalid UTF-8 sequence");
         return -1;
       }
-      return handle(hdr.opcode, payload, frame_size);
+      // Call upper layer.
+      validation_offset_ = 0;
+      return handle(hdr_.opcode, payload, hdr_.payload_len);
     }
-    if (hdr.opcode != detail::rfc6455::continuation_frame) {
+    if (hdr_.opcode != detail::rfc6455::opcode_type::continuation_frame) {
       CAF_LOG_DEBUG("expected a WebSocket continuation_frame");
       abort_and_shutdown(sec::protocol_error,
                          "expected a WebSocket continuation_frame");
       return -1;
     }
     // End of fragmented input.
-    payload_buf_.insert(payload_buf_.end(), payload.begin(), payload.end());
-    if (opcode_ == detail::rfc6455::text_frame && !payload_valid()) {
+    if (opcode_ != detail::rfc6455::opcode_type::text_frame) {
+      payload_buf_.insert(payload_buf_.end(), payload.begin(), payload.end());
+    }
+    if (opcode_ == detail::rfc6455::opcode_type::text_frame
+        && validation_offset_ != payload_buf_.size()) {
       abort_and_shutdown(sec::malformed_message, "invalid UTF-8 sequence");
       return -1;
     }
-    auto result = handle(opcode_, payload_buf_, frame_size);
-    opcode_ = nil_code;
+    auto result = handle(opcode_, payload_buf_, hdr_.payload_len);
+    opcode_ = detail::rfc6455::opcode_type::nil_code;
     payload_buf_.clear();
     validation_offset_ = 0;
     return result;
   }
   // The first frame must not be a continuation frame. Any frame that is not
   // the first frame must be a continuation frame.
-  if (opcode_ == nil_code) {
-    opcode_ = hdr.opcode;
-  } else if (hdr.opcode != detail::rfc6455::continuation_frame) {
+  if (opcode_ == detail::rfc6455::opcode_type::nil_code) {
+    opcode_ = hdr_.opcode;
+  } else if (hdr_.opcode != detail::rfc6455::opcode_type::continuation_frame) {
     CAF_LOG_DEBUG("expected a continuation frame");
     abort_and_shutdown(sec::protocol_error, "expected a continuation frame");
     return -1;
   }
-  payload_buf_.insert(payload_buf_.end(), payload.begin(), payload.end());
-  if (opcode_ == detail::rfc6455::text_frame && !payload_valid()) {
-    abort_and_shutdown(sec::malformed_message, "invalid UTF-8 sequence");
-    return -1;
+  if (opcode_ != detail::rfc6455::opcode_type::text_frame) {
+    payload_buf_.insert(payload_buf_.end(), payload.begin(), payload.end());
+  }
+  auto ret = static_cast<ptrdiff_t>(hdr_.payload_len);
+  hdr_.opcode = detail::rfc6455::opcode_type::nil_code;
+  return ret;
+}
+
+ptrdiff_t framing::consume(byte_span buffer, byte_span delta) {
+  // Make sure we're overriding any 'exactly' setting.
+  down_->configure_read(receive_policy::up_to(2048));
+  if (!hdr_) {
```

Comment:
As per previous suggestion:

```suggestion
  if (!hdr_.valid()) {
```

Reply:
:+1: